layers/cmd/schema.go

```diff
@@ -15,20 +15,252 @@
 	"fmt"
+	"io"
+	"io/ioutil"
 	"path/filepath"
+	"strings"
+	"time"
 
 	"github.com/cloudprivacylabs/lsa/layers/cmd/cmdutil"
+	jsonsch "github.com/cloudprivacylabs/lsa/pkg/json"
 	"github.com/cloudprivacylabs/lsa/pkg/ls"
+	"github.com/santhosh-tekuri/jsonschema/v5"
 )
 
-type SchemaOverlays struct {
-	Schema   string   `json:"schema"`
-	Overlays []string `json:"overlays"`
+// type SchemaOverlays[T OverlayTypeConstraint] struct {
+// 	Schema     string     `json:"schema"`
+// 	JSONSchema JSONSchema `json:"jsonSchema"`
+// 	Overlays   []T        `json:"overlays"`
+// }
+
+// type JSONSchema struct {
+// 	Schema  string `json:"schema"`
+// 	LayerId string `json:"layerId"`
+// }
+
+// type OverlayTypeConstraint interface {
+// 	string | JSONSchema
+// }
+
+// --- >
+// Bundle defines type names for variants so references can be resolved
+type Bundle struct {
+	ID        string                   `json:"id" bson:"id" yaml:"id" bson:"id"`
+	CreatedAt time.Time                `json:"createdAt" bson:"createdAt" yaml:"createdAt"`
+	UpdatedAt time.Time                `json:"updatedAt" bson:"updatedAt" yaml:"updatedAt"`
+	Variants  map[string]BundleVariant `json:"variants" yaml:"variants"`
+}
+
+type BundleSchemaRef struct {
+	LayerID    string               `json:"layerId,omitempty" yaml:"layerId,omitempty" bson:"layerId,omitempty"`
+	JSONSchema *JSONSchemaReference `json:"jsonSchema" yaml:"jsonSchema" bson:"jsonSchema,omitempty"`
+}
+
+// GetLayerID returns the layer id for the schema reference
+func (ref BundleSchemaRef) GetLayerID() string {
+	if ref.JSONSchema != nil {
+		return ref.JSONSchema.LayerID
+	}
+	return ref.LayerID
+}
+
+type JSONSchemaReference struct {
+	LayerID string `json:"layerId" yaml:"layerId" bson:"layerId"`
+	Ref     string `json:"ref" yaml:"ref" bson:"ref"`
+}
+
+// BundleVariant combines a schema and overlays
+type BundleVariant struct {
+	BundleSchemaRef
+	Overlays []BundleSchemaRef `json:"overlays" yaml:"overlays"`
+}
+
+// ParseBundle parses a bundle from JSON or YAML input
+func ParseBundle(text string, contentType string) (*Bundle, error) {
+	var ret Bundle
+	if err := cmdutil.ReadJSON(contentType, &ret); err != nil {
+		return nil, err
+	}
+	return &ret, nil
+}
+
+func (bundle *Bundle) importJSONSchema(ctx context.Context, loader ls.SchemaLoader, importEntities []jsonsch.Entity) (map[string]*ls.Layer, error) {
+	compiler := jsonschema.NewCompiler()
+	compiler.LoadURL = func(s string) (io.ReadCloser, error) {
+		obj, err := loader.LoadSchema(s)
+		if err != nil {
+			return nil, err
+		}
+		return ioutil.NopCloser(strings.NewReader(obj.GetID())), nil
+	}
+	// Import all JSON schemas into a graph
+	jsonLayers := make(map[string]*ls.Layer)
+	if len(importEntities) > 0 {
+		compiled, err := jsonsch.CompileEntitiesWith(compiler, importEntities...)
+		if err != nil {
+			return nil, err
+		}
+		g := ls.NewLayerGraph()
+		layers, err := jsonsch.BuildEntityGraph(g, ls.SchemaTerm, jsonsch.LinkRefsByValueType, compiled...)
```

Comment:
this part is wrong. It uses ls.SchemaTerm. Get the term as an arg to the method. This should be ls.SchemaTerm for schemas, ls.OverlayTerm for overlays.